File: Labs/Lab5/Lab5.java
public class TreeNode {
     int val;
     TreeNode left;
     TreeNode right;

     TreeNode() {}
     TreeNode(int val) { this.val = val; }
     TreeNode(int val, TreeNode left, TreeNode right) {

         this.val = val;
         this.left = left;
         this.right = right;
     }
  }
public class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean isValidBST(TreeNode node, long min, long max) {
        if (node == null) {
            return true;
        }

        if (node.val <= min || node.val >= max) {
            return false;
        }

        return isValidBST(node.left, min, node.val) && isValidBST(node.right, node.val, max);
    }
}



// We define a TreeNode class representing the nodes of the binary tree.
// Create a Solution class containing a public method `isValidBST` that takes the root node of the binary tree as input and returns a boolean indicating the tree is a valid BST.
// At each node, it checks if the node's value is within the valid range defined by `min` and `max`. If the node's value violates this range, the method returns false, that the tree is not a valid BST.
// If the current node is null, the method returns true.
// Use `long` type for `min` and `max` parameters to handle edge cases where node values might be at the limits of integer range.

File: Labs/Lab5/Lab5.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def isValidBST(self, root):
        prev_node = None
        stack = []
        
        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            
            root = stack.pop()
            
            if prev_node is not None and root.val <= prev_node.val:
                return False
            
            prev_node = root
            
            root = root.right
        
        return True



# We define a TreeNode class to represent the nodes of the binary tree, similar to the Java implementation.
# Public method `isValidBST' takes the root of the binary tree as input and returns a boolean whether the tree is a valid BST. 
# At each node, it checks if the node's value is within the valid range defined by `min_val` and `max_val`. If the node's value not in this range, the method returns false, indicating the tree is not a valid BST. Otherwise, it calls itself for the left and right subtrees, updating the `min_val` and `max_val` values.
# If the current node is None indicating the end of a branch, the method returns true.
